chain/indexer/src/streamer/streamer.rs

```diff
@@ -84,39 +92,64 @@ async fn fetch_block_by_height(
         .map_err(|err| FailedToFetchData::String(err))
 }
 
-/// This function supposed to return the entire `BlockResponse`.
-/// It fetches the block and fetches all the chunks for the block
+/// This function supposed to return the entire `StreamerMessage`.
+/// It fetches the block and all related parts (chunks, outcomes, state changes etc.)
 /// and returns everything together in one struct
-async fn fetch_block_response(
+async fn build_streamer_message(
     client: &Addr<near_client::ViewClientActor>,
     block: views::BlockView,
-    outcomes_to_get: impl Iterator<Item = types::TransactionOrReceiptId>,
-) -> Result<FetchBlockResponse, FailedToFetchData> {
+) -> Result<StreamerMessage, FailedToFetchData> {
     let chunks = fetch_chunks(&client, &block.chunks).await?;
-    let (outcomes, mut outcomes_to_retry) = fetch_outcomes(&client, outcomes_to_get).await;
 
-    for chunk in &chunks {
-        outcomes_to_retry.extend(chunk.transactions.iter().map(|transaction| {
-            types::TransactionOrReceiptId::Transaction {
-                transaction_hash: transaction.hash.clone(),
-                sender_id: transaction.signer_id.to_string(),
-            }
-        }));
+    let mut local_receipts: Vec<views::ReceiptView> = vec![];
+    let mut outcomes = fetch_outcomes(&client, block.header.hash).await?;
+    let mut indexer_chunks: Vec<IndexerChunkView> = vec![];
 
-        outcomes_to_retry.extend(chunk.receipts.iter().map(|receipt| {
-            types::TransactionOrReceiptId::Receipt {
-                receipt_id: receipt.receipt_id,
-                receiver_id: receipt.receiver_id.to_string(),
-            }
-        }));
+    for chunk in &chunks {
+        let indexer_transactions = chunk
+            .transactions
+            .iter()
+            .map(|transaction| {
+                if transaction.signer_id == transaction.receiver_id {
+                    // This transaction generates local receipt (sir)
+                    local_receipts.push(views::ReceiptView {
+                        predecessor_id: transaction.signer_id.clone(),
+                        receiver_id: transaction.receiver_id.clone(),
+                        receipt_id: outcomes
+                            .get(&transaction.hash)
+                            .unwrap()
```

Comment:
If you construct `indexer_transactions` first and then `.filter.map` them as suggested above, you won't need to query the outcome twice and do `.unwrap` (ideally, we should express everything with as little unwraps as possible to aboid lurking bugs.

P.S. Vec has `.extend` method that accepts an iterator.